File: orb/apps/ios/kagami-ios/KagamiIOS/Services/EarconService.swift
//
// EarconService.swift — BBC Symphony Orchestra Earcons for iOS
//
// Colony: Nexus (e₄) — Integration
//
// Features:
//   - 36 BBC Symphony Orchestra earcons (REAPER + BBC SO VST)
//   - Tier 1 earcons bundled with app for instant playback
//   - Tier 2 earcons lazy-loaded from CDN
//   - Integrated with KagamiHaptics for coordinated feedback
//   - Spatial audio support via AVAudioEngine
//
// Architecture:
//   Event → EarconService → AVAudioEngine → Spatial Output
//                        → KagamiHaptics → Taptic Engine
//
// Created: January 12, 2026
// 鏡

import AVFoundation
import Combine

/// Service for playing BBC Symphony Orchestra earcons
@MainActor
final class EarconService: ObservableObject {

    // MARK: - Singleton

    static let shared = EarconService()

    // MARK: - Published State

    @Published var isInitialized = false
    @Published var isMuted = false
    @Published var masterVolume: Float = 0.7
    @Published var tier1Loaded = false

    // MARK: - Types

    /// Available earcons (36 total)
    enum Earcon: String, CaseIterable {
        // Tier 1 — Core (14 bundled)
        case notification
        case success
        case error
        case alert
        case arrival
        case departure
        case celebration
        case settling
        case awakening
        case cinematic
        case focus
        case securityArm = "security_arm"
        case package
        case meetingSoon = "meeting_soon"

        // Tier 2 — Extended (22 CDN)
        case roomEnter = "room_enter"
        case doorOpen = "door_open"
        case doorClose = "door_close"
        case lockEngaged = "lock_engaged"
        case voiceAcknowledge = "voice_acknowledge"
        case voiceComplete = "voice_complete"
        case washerComplete = "washer_complete"
        case coffeeReady = "coffee_ready"
        case morningSequence = "morning_sequence"
        case eveningTransition = "evening_transition"
        case midnight
        case stormApproaching = "storm_approaching"
        case rainStarting = "rain_starting"
        case motionDetected = "motion_detected"
        case cameraAlert = "camera_alert"
        case messageReceived = "message_received"
        case homeEmpty = "home_empty"
        case firstHome = "first_home"
        case ovenPreheat = "oven_preheat"
        case dishwasherComplete = "dishwasher_complete"
        case dryerComplete = "dryer_complete"

        /// Whether this earcon is bundled (Tier 1)
        var isTier1: Bool {
            switch self {
            case .notification, .success, .error, .alert, .arrival, .departure,
                 .celebration, .settling, .awakening, .cinematic, .focus,
                 .securityArm, .package, .meetingSoon:
                return true
            default:
                return false
            }
        }

        /// Corresponding haptic pattern
        var hapticPattern: HapticPattern {
            switch self {
            case .success, .celebration, .arrival:
                return .success
            case .error:
                return .error
            case .alert, .cameraAlert:
                return .warning
            case .notification, .messageReceived, .package:
                return .mediumImpact
            case .focus, .roomEnter:
                return .lightImpact
            case .lockEngaged, .securityArm:
                return .heavyImpact
            case .departure, .settling, .homeEmpty:
                return .softImpact
            case .voiceAcknowledge:
                return .selection
            case .voiceComplete:
                return .success
            default:
                return .lightImpact
            }
        }
    }

    /// Semantic feedback events that map to earcons
    enum FeedbackEvent: String {
        case tap
        case select
        case sceneActivated
        case lightsOn
        case lightsOff
        case shadeOpen
        case shadeClose

        var earcon: Earcon {
            switch self {
            case .tap, .select:
                return .focus
            case .sceneActivated:
                return .success
            case .lightsOn, .shadeOpen:
                return .doorOpen
            case .lightsOff, .shadeClose:
                return .doorClose
            }
        }
    }

    // MARK: - Private State

    private var audioEngine: AVAudioEngine?
    private var playerNode: AVAudioPlayerNode?
    private var mixerNode: AVAudioMixerNode?
    private var environmentNode: AVAudioEnvironmentNode?

    /// Cached audio buffers
    private var bufferCache: [String: AVAudioPCMBuffer] = [:]

    /// Loading tasks
    private var loadingTasks: [String: Task<Bool, Never>] = [:]

    /// CDN base URL
    private let cdnBaseURL = "https://storage.googleapis.com/kagami-media-public/earcons/v1/aac"

    // MARK: - Init

    private init() {
        Task {
            await initialize()
        }
    }

    // MARK: - Initialization

    func initialize() async {
        guard !isInitialized else { return }

        do {
            // Configure audio session
            let session = AVAudioSession.sharedInstance()
            try session.setCategory(.playback, mode: .default, options: [.mixWithOthers])
            try session.setActive(true)

            // Setup audio engine
            audioEngine = AVAudioEngine()
            playerNode = AVAudioPlayerNode()
            environmentNode = AVAudioEnvironmentNode()

            guard let engine = audioEngine,
                  let player = playerNode,
                  let environment = environmentNode else {
                print("EarconService: Failed to create audio nodes")
                return
            }

            engine.attach(player)
            engine.attach(environment)

            // Connect: player -> environment -> main output
            let format = AVAudioFormat(standardFormatWithSampleRate: 48000, channels: 2)!
            engine.connect(player, to: environment, format: format)
            engine.connect(environment, to: engine.mainMixerNode, format: format)

            // Configure spatial environment
            environment.listenerPosition = AVAudio3DPoint(x: 0, y: 0, z: 0)
            environment.renderingAlgorithm = .auto

            try engine.start()

            isInitialized = true
            print("EarconService: Initialized")

            // Preload Tier 1 earcons
            await preloadTier1()

        } catch {
            print("EarconService: Initialization failed - \(error)")
        }
    }

    /// Preload all Tier 1 earcons from bundle
    func preloadTier1() async {
        let tier1Earcons = Earcon.allCases.filter { $0.isTier1 }

        var loadedCount = 0
        for earcon in tier1Earcons {
            if await loadFromBundle(earcon.rawValue) {
                loadedCount += 1
            }
        }

        tier1Loaded = loadedCount > 0
        print("EarconService: Loaded \(loadedCount)/\(tier1Earcons.count) Tier 1 earcons")
    }

    // MARK: - Playback

    /// Play an earcon with optional haptic feedback
    /// - Parameters:
    ///   - earcon: The earcon to play
    ///   - withHaptic: Whether to also trigger haptic feedback
    ///   - position: Optional 3D position for spatial audio
    func play(_ earcon: Earcon, withHaptic: Bool = true, position: AVAudio3DPoint? = nil) {
        guard isInitialized, !isMuted else { return }

        // Play haptic first (lower latency)
        if withHaptic {
            KagamiHaptics.shared.play(earcon.hapticPattern)
        }

        // Play audio
        Task {
            await playAudio(earcon.rawValue, at: position)
        }
    }

    /// Play a semantic feedback event
    func play(_ event: FeedbackEvent, withHaptic: Bool = true) {
        play(event.earcon, withHaptic: withHaptic)
    }

    /// Play an earcon by name (for dynamic usage)
    func play(named name: String, withHaptic: Bool = true) {
        guard let earcon = Earcon(rawValue: name) else {
            print("EarconService: Unknown earcon: \(name)")
            return
        }
        play(earcon, withHaptic: withHaptic)
    }

    // MARK: - Audio Playback

    private func playAudio(_ name: String, at position: AVAudio3DPoint?) async {
        guard let player = playerNode, let environment = environmentNode else { return }

        // Get or load buffer
        let buffer: AVAudioPCMBuffer?
        if let cached = bufferCache[name] {
            buffer = cached
        } else {
            // Try bundle first, then CDN
            var loaded = await loadFromBundle(name)
            if !loaded {
                loaded = await loadFromCDN(name)
            }
            if loaded {
                buffer = bufferCache[name]
            } else {
                print("EarconService: Failed to load earcon: \(name)")
                return
            }
        }

        guard let audioBuffer = buffer else { return }

        // Set spatial position
        if let pos = position {
            environment.listenerPosition = pos
        }

        // Set volume
        player.volume = masterVolume

        // Stop any current playback
        if player.isPlaying {
            player.stop()
        }

        // Play
        await player.scheduleBuffer(audioBuffer, at: nil, options: .interrupts)
        player.play()
    }

    // MARK: - Loading

    private func loadFromBundle(_ name: String) async -> Bool {
        guard bufferCache[name] == nil else { return true }

        guard let url = Bundle.main.url(forResource: name, withExtension: "m4a", subdirectory: "Earcons") else {
            return false
        }

        do {
            let file = try AVAudioFile(forReading: url)
            let format = file.processingFormat
            let frameCount = AVAudioFrameCount(file.length)

            guard let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frameCount) else {
                return false
            }

            try file.read(into: buffer)
            bufferCache[name] = buffer

            print("EarconService: Loaded from bundle: \(name)")
            return true

        } catch {
            print("EarconService: Bundle load failed for \(name): \(error)")
            return false
        }
    }

    private func loadFromCDN(_ name: String) async -> Bool {
        guard bufferCache[name] == nil else { return true }

        // Check for existing task
        if let existingTask = loadingTasks[name] {
            return await existingTask.value
        }

        let task = Task<Bool, Never> {
            guard let url = URL(string: "\(cdnBaseURL)/\(name).m4a") else {
                print("EarconService: Invalid CDN URL for \(name)")
                return false
            }

            do {
                let (data, response) = try await URLSession.shared.data(from: url)

                guard let httpResponse = response as? HTTPURLResponse,
                      httpResponse.statusCode == 200 else {
                    return false
                }

                // Write to temp file
                let tempURL = FileManager.default.temporaryDirectory.appendingPathComponent("\(name).m4a")
                try data.write(to: tempURL)

                let file = try AVAudioFile(forReading: tempURL)
                let format = file.processingFormat
                let frameCount = AVAudioFrameCount(file.length)

                guard let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frameCount) else {
                    return false
                }

                try file.read(into: buffer)
                bufferCache[name] = buffer

                // Clean up
                try? FileManager.default.removeItem(at: tempURL)

                print("EarconService: Downloaded from CDN: \(name)")
                return true

            } catch {
                print("EarconService: CDN download failed for \(name): \(error)")
                return false
            }
        }

        loadingTasks[name] = task
        let result = await task.value
        loadingTasks.removeValue(forKey: name)
        return result
    }

    // MARK: - Volume Control

    func setVolume(_ volume: Float) {
        masterVolume = max(0, min(1, volume))
    }

    func mute() {
        isMuted = true
    }

    func unmute() {
        isMuted = false
    }

    // MARK: - Utility

    /// Check if an earcon is loaded
    func isLoaded(_ earcon: Earcon) -> Bool {
        bufferCache[earcon.rawValue] != nil
    }

    /// Preload specific earcons
    func preload(_ earcons: [Earcon]) async {
        for earcon in earcons {
            if !bufferCache.keys.contains(earcon.rawValue) {
                _ = earcon.isTier1
                    ? await loadFromBundle(earcon.rawValue)
                    : await loadFromCDN(earcon.rawValue)
            }
        }
    }

    /// Clear cache (for memory pressure)
    func clearCache() {
        let tier1Names = Set(Earcon.allCases.filter { $0.isTier1 }.map { $0.rawValue })
        // Keep Tier 1, clear Tier 2
        bufferCache = bufferCache.filter { tier1Names.contains($0.key) }
    }
}

/*
 * 鏡
 * h(x) >= 0. Always.
 *
 * BBC Symphony Orchestra earcons provide virtuoso audio feedback.
 * Each earcon is a complete musical phrase with emotional intent.
 * Coordinated with haptics for multi-sensory experience.
 */

File: orb/apps/android/kagami-android/app/src/main/java/com/kagami/android/services/EarconService.kt
package com.kagami.android.services

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.MediaPlayer
import android.net.Uri
import android.os.Build
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.*
import java.io.File
import java.net.URL
import javax.inject.Inject
import javax.inject.Singleton

/**
 * EarconService — BBC Symphony Orchestra Earcons for Android
 *
 * Features:
 * - 36 BBC Symphony Orchestra earcons (REAPER + BBC SO VST)
 * - Tier 1 earcons bundled with app in assets/earcons/
 * - Tier 2 earcons lazy-loaded from CDN
 * - Integrated with KagamiHaptics for coordinated feedback
 * - ExoPlayer for high-quality playback
 *
 * Architecture:
 *   Event → EarconService → MediaPlayer → Audio Output
 *                        → KagamiHaptics → Vibration
 *
 * h(x) >= 0. Always.
 */

/**
 * Available earcons (36 total)
 */
enum class Earcon(val fileName: String) {
    // Tier 1 — Core (14 bundled)
    NOTIFICATION("notification"),
    SUCCESS("success"),
    ERROR("error"),
    ALERT("alert"),
    ARRIVAL("arrival"),
    DEPARTURE("departure"),
    CELEBRATION("celebration"),
    SETTLING("settling"),
    AWAKENING("awakening"),
    CINEMATIC("cinematic"),
    FOCUS("focus"),
    SECURITY_ARM("security_arm"),
    PACKAGE("package"),
    MEETING_SOON("meeting_soon"),

    // Tier 2 — Extended (22 CDN)
    ROOM_ENTER("room_enter"),
    DOOR_OPEN("door_open"),
    DOOR_CLOSE("door_close"),
    LOCK_ENGAGED("lock_engaged"),
    VOICE_ACKNOWLEDGE("voice_acknowledge"),
    VOICE_COMPLETE("voice_complete"),
    WASHER_COMPLETE("washer_complete"),
    COFFEE_READY("coffee_ready"),
    MORNING_SEQUENCE("morning_sequence"),
    EVENING_TRANSITION("evening_transition"),
    MIDNIGHT("midnight"),
    STORM_APPROACHING("storm_approaching"),
    RAIN_STARTING("rain_starting"),
    MOTION_DETECTED("motion_detected"),
    CAMERA_ALERT("camera_alert"),
    MESSAGE_RECEIVED("message_received"),
    HOME_EMPTY("home_empty"),
    FIRST_HOME("first_home"),
    OVEN_PREHEAT("oven_preheat"),
    DISHWASHER_COMPLETE("dishwasher_complete"),
    DRYER_COMPLETE("dryer_complete");

    /** Whether this earcon is bundled (Tier 1) */
    val isTier1: Boolean
        get() = when (this) {
            NOTIFICATION, SUCCESS, ERROR, ALERT, ARRIVAL, DEPARTURE,
            CELEBRATION, SETTLING, AWAKENING, CINEMATIC, FOCUS,
            SECURITY_ARM, PACKAGE, MEETING_SOON -> true
            else -> false
        }

    /** Corresponding haptic pattern */
    val hapticPattern: HapticPattern
        get() = when (this) {
            SUCCESS, CELEBRATION, ARRIVAL -> HapticPattern.SUCCESS
            ERROR -> HapticPattern.ERROR
            ALERT, CAMERA_ALERT -> HapticPattern.WARNING
            NOTIFICATION, MESSAGE_RECEIVED, PACKAGE -> HapticPattern.MEDIUM_IMPACT
            FOCUS, ROOM_ENTER -> HapticPattern.LIGHT_IMPACT
            LOCK_ENGAGED, SECURITY_ARM -> HapticPattern.HEAVY_IMPACT
            DEPARTURE, SETTLING, HOME_EMPTY -> HapticPattern.SOFT_IMPACT
            VOICE_ACKNOWLEDGE -> HapticPattern.SELECTION
            VOICE_COMPLETE -> HapticPattern.SUCCESS
            else -> HapticPattern.LIGHT_IMPACT
        }

    companion object {
        fun fromFileName(name: String): Earcon? = entries.find { it.fileName == name }
    }
}

/**
 * Semantic feedback events that map to earcons
 */
enum class FeedbackEvent(val earcon: Earcon) {
    TAP(Earcon.FOCUS),
    SELECT(Earcon.FOCUS),
    SCENE_ACTIVATED(Earcon.SUCCESS),
    LIGHTS_ON(Earcon.DOOR_OPEN),
    LIGHTS_OFF(Earcon.DOOR_CLOSE),
    SHADE_OPEN(Earcon.DOOR_OPEN),
    SHADE_CLOSE(Earcon.DOOR_CLOSE),
}

@Singleton
class EarconService @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val mainScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    // Audio
    private var mediaPlayer: MediaPlayer? = null
    private val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager

    // Caching
    private val cacheDir = File(context.cacheDir, "earcons")
    private val loadingJobs = mutableMapOf<String, Job>()

    // State
    var isInitialized = false
        private set
    var isMuted = false
    var masterVolume = 0.7f
        set(value) {
            field = value.coerceIn(0f, 1f)
            mediaPlayer?.setVolume(field, field)
        }
    var tier1Loaded = false
        private set

    // Haptics integration
    private val haptics: KagamiHaptics by lazy { KagamiHaptics.getInstance(context) }

    // CDN
    private val cdnBaseUrl = "https://storage.googleapis.com/kagami-media-public/earcons/v1/mp3"

    init {
        initialize()
    }

    /**
     * Initialize the earcon service
     */
    fun initialize() {
        if (isInitialized) return

        // Create cache directory
        cacheDir.mkdirs()

        // Setup media player
        mediaPlayer = MediaPlayer().apply {
            setAudioAttributes(
                AudioAttributes.Builder()
                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                    .setUsage(AudioAttributes.USAGE_NOTIFICATION_EVENT)
                    .build()
            )
            setVolume(masterVolume, masterVolume)
        }

        isInitialized = true

        // Preload Tier 1 earcons
        scope.launch {
            preloadTier1()
        }
    }

    /**
     * Preload all Tier 1 earcons from assets
     */
    private suspend fun preloadTier1() {
        val tier1Earcons = Earcon.entries.filter { it.isTier1 }
        var loadedCount = 0

        for (earcon in tier1Earcons) {
            if (loadFromAssets(earcon.fileName)) {
                loadedCount++
            }
        }

        tier1Loaded = loadedCount > 0
        println("EarconService: Loaded $loadedCount/${tier1Earcons.size} Tier 1 earcons")
    }

    // ========================================================================
    // Playback
    // ========================================================================

    /**
     * Play an earcon with optional haptic feedback
     */
    fun play(
        earcon: Earcon,
        withHaptic: Boolean = true,
    ) {
        if (!isInitialized || isMuted) return

        // Play haptic first (lower latency)
        if (withHaptic) {
            haptics.play(earcon.hapticPattern)
        }

        // Play audio
        scope.launch {
            playAudio(earcon.fileName)
        }
    }

    /**
     * Play a semantic feedback event
     */
    fun play(event: FeedbackEvent, withHaptic: Boolean = true) {
        play(event.earcon, withHaptic)
    }

    /**
     * Play an earcon by name (for dynamic usage)
     */
    fun playByName(name: String, withHaptic: Boolean = true) {
        val earcon = Earcon.fromFileName(name)
        if (earcon != null) {
            play(earcon, withHaptic)
        } else {
            println("EarconService: Unknown earcon: $name")
        }
    }

    private suspend fun playAudio(name: String) {
        val file = getCachedFile(name)

        if (file?.exists() == true) {
            playFile(file)
        } else {
            // Try to load
            val loaded = if (Earcon.fromFileName(name)?.isTier1 == true) {
                loadFromAssets(name)
            } else {
                loadFromCDN(name)
            }

            if (loaded) {
                getCachedFile(name)?.let { playFile(it) }
            }
        }
    }

    private fun playFile(file: File) {
        mainScope.launch {
            try {
                mediaPlayer?.apply {
                    reset()
                    setDataSource(file.absolutePath)
                    prepare()
                    setVolume(masterVolume, masterVolume)
                    start()
                }
            } catch (e: Exception) {
                println("EarconService: Playback failed: ${e.message}")
            }
        }
    }

    // ========================================================================
    // Loading
    // ========================================================================

    private fun getCachedFile(name: String): File? {
        val file = File(cacheDir, "$name.mp3")
        return if (file.exists()) file else null
    }

    private suspend fun loadFromAssets(name: String): Boolean {
        val cacheFile = File(cacheDir, "$name.mp3")
        if (cacheFile.exists()) return true

        return try {
            context.assets.open("earcons/$name.mp3").use { input ->
                cacheFile.outputStream().use { output ->
                    input.copyTo(output)
                }
            }
            true
        } catch (e: Exception) {
            // Asset not found, will try CDN
            false
        }
    }

    private suspend fun loadFromCDN(name: String): Boolean {
        val cacheFile = File(cacheDir, "$name.mp3")
        if (cacheFile.exists()) return true

        // Check for existing loading job
        if (loadingJobs.containsKey(name)) {
            loadingJobs[name]?.join()
            return cacheFile.exists()
        }

        val job = scope.launch {
            try {
                val url = URL("$cdnBaseUrl/$name.mp3")
                val connection = url.openConnection()
                connection.connectTimeout = 10000
                connection.readTimeout = 30000

                connection.getInputStream().use { input ->
                    cacheFile.outputStream().use { output ->
                        input.copyTo(output)
                    }
                }

                println("EarconService: Downloaded from CDN: $name")

            } catch (e: Exception) {
                println("EarconService: CDN download failed for $name: ${e.message}")
                cacheFile.delete()
            }
        }

        loadingJobs[name] = job
        job.join()
        loadingJobs.remove(name)

        return cacheFile.exists()
    }

    // ========================================================================
    // Controls
    // ========================================================================

    fun mute() {
        isMuted = true
    }

    fun unmute() {
        isMuted = false
    }

    /**
     * Check if an earcon is loaded
     */
    fun isLoaded(earcon: Earcon): Boolean {
        return getCachedFile(earcon.fileName)?.exists() == true
    }

    /**
     * Preload specific earcons
     */
    fun preload(earcons: List<Earcon>) {
        scope.launch {
            for (earcon in earcons) {
                if (!isLoaded(earcon)) {
                    if (earcon.isTier1) {
                        loadFromAssets(earcon.fileName)
                    } else {
                        loadFromCDN(earcon.fileName)
                    }
                }
            }
        }
    }

    /**
     * Clear Tier 2 cache (keep Tier 1)
     */
    fun clearTier2Cache() {
        cacheDir.listFiles()?.forEach { file ->
            val name = file.nameWithoutExtension
            if (Earcon.fromFileName(name)?.isTier1 == false) {
                file.delete()
            }
        }
    }

    /**
     * Get total cache size in bytes
     */
    fun getCacheSize(): Long {
        return cacheDir.listFiles()?.sumOf { it.length() } ?: 0
    }

    /**
     * Release resources
     */
    fun release() {
        mediaPlayer?.release()
        mediaPlayer = null
        scope.cancel()
        mainScope.cancel()
        isInitialized = false
    }

    companion object {
        @Volatile
        private var instance: EarconService? = null

        fun getInstance(context: Context): EarconService {
            return instance ?: synchronized(this) {
                instance ?: EarconService(context.applicationContext).also { instance = it }
            }
        }
    }
}

/*
 * 鏡
 * h(x) >= 0. Always.
 *
 * BBC Symphony Orchestra earcons provide virtuoso audio feedback.
 * Each earcon is a complete musical phrase with emotional intent.
 * Coordinated with haptics for multi-sensory experience.
 */
